### PR TITLE
HRNR

### PR DESCRIPTION
**Pull request by PotassiumWings**
初版。目前没有做成固定下游任务、可改上游的形式，预训练两个转移矩阵的训练过程非常神秘，梯度不怎么变化。

目前实现和论文里的不同之处：

1. 计算 TRZ M2 用的 GCN 而不是 GAT。（GAT 做不太出来，矩阵只有十几个位置 > 0.5）
2. 计算 TRZ loss 中，C 矩阵直接用了可达矩阵模拟，没有使用真实路段数据。

TODO 的东西比较多：

1. 预训练的训练次数、学习率做成参数；
2. 使用 osm extract 路网制作出原子数据，然后进行比较；
3. 可达矩阵改成论文中所写，真实下一跳数据计算出来的 C；
4. 论文实现的后半部分照着原文实现一下；
5. 增加数据缓存机制；
6. evaluator 还没写，目前仅看训练的效果也不是很好。

固定上下游任务有点困难，整个模型从设计来说就是一体的，除非违背论文设计 NS 的构造：让 NS 表示地更广一些，不仅仅是那四个 feature。这个也待讨论。

### CODE REVIEW

File: libcity/model/road_representation/HRNR.py [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+
+        self.linear_red_dim = torch.nn.Linear(hparams.hidden_dims, 100)
+        self.node_emb, self.init_emb = None, None
+
+    def forward(self, input_bat):  # batch_size * length * dims
+        self.node_emb = self.graph_enc(self.node_feature, self.type_feature, self.length_feature, self.lane_feature,
+                                       self.adj)
+        self.init_emb = self.graph_enc.init_feat
+        output_state = torch.cat((self.node_emb[input_bat], self.init_emb[input_bat]), 1)
+        pred_tra = self.linear(output_state)
```
Comment (WenMellors) [inline libcity/model/road_representation/HRNR.py:46]: 你这里可以不过这个 Linear 层，直接返回训练好的全部路段对应的表征矩阵吗？也就是 Ns

Comment (WenMellors) [reply libcity/model/road_representation/HRNR.py:46]: 这样你应该就可以实现下游任务的切换了吧

**Review by WenMellors — CHANGES_REQUESTED** [review]

File: libcity/model/road_representation/HRNR.py (hunk 1 above)
Comment (WenMellors) [reply libcity/model/road_representation/HRNR.py:46]: 然后下游任务的模型接纳你这个 Ns 来做自回归或者标签预测

Comment (PotassiumWings) [reply libcity/model/road_representation/HRNR.py:46]: 问题在于 dataset 也要改

Comment (PotassiumWings) [reply libcity/model/road_representation/HRNR.py:46]: 也就是需要两个 dataset（一个是路网表征所需要的，一个是下游任务的数据格式）？

File: libcity/executor/HRNR_executor.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        max_f1 = 0
+        max_auc = 0
+        count = 0
+        model_optimizer = torch.optim.Adam(self.model.parameters(), lr=hparams.lp_learning_rate)
+        eval_dataloader_iter = iter(eval_dataloader)
+        for i in range(hparams.label_epoch):
+            self._logger.info("epoch " + str(i) + ", processed " + str(count))
+            for step, (train_set, train_label) in enumerate(train_dataloader):
+                model_optimizer.zero_grad()
+                train_set = train_set.clone().detach()
```
Comment (WenMellors) [inline libcity/executor/HRNR_executor.py:30]: 所以这里如果不做 clone 的话，下一轮次 train_set 就会带有上一轮次反向传播过来的梯度？

File: libcity/model/road_representation/HRNR.py (hunk 1 above)
Comment (WenMellors) [reply libcity/model/road_representation/HRNR.py:46]: 那确实是需要的。那你觉得有必要把上游任务和下游任务区解耦开来吗？

**Comment by aptx1231** [comment]
已线下沟通

**Comment by WenMellors** [comment]
路网表征模型的接口类，两个下游任务的执行器并配套下游任务的模型。
* 这个下游模型可以放到执行器 py 里面，或者放到 `model` 文件夹下

**Comment by aptx1231** [comment]
没跑吧？

**Comment by aptx1231** [comment]
重开分支，重新PR吧，这个分支也Merge不了
